### PR TITLE
fs: Filesystem module (WIP)

### PR DESCRIPTION
**Pull request by authmillenon**
This PR creates a common interface for all filesystems to use.

STILL WIP

### CODE REVIEW

**Comment by gebart** [comment]
Do we have any candidates for implementations of filesystems?

Contiki has CFS, but that is BSD licensed, so we can't use that code.

There quite a lot of different file systems for larger devices (from several megabytes and up), but what other filesystems exist for tiny (most likely flash-) devices (from a few kilobytes up to tens of megabytes)?

**Comment by authmillenon** [comment]
The original idea was to add CFS as a [package](https://github.com/RIOT-OS/RIOT/tree/master/pkg). This way we would only include patches, if need be but not the code. I also discovered and thought recently about FAT via [FatFS](http://elm-chan.org/fsw/ff/00index_e.html).

On continueing this PR: Hope I get some time beside the network stack to fix the documentation. As far as I remember, this was the only thing lacking in this PR.

**Comment by gebart** [comment]
I only briefly looked over FatFS now. The license page is somewhat confusingly worded, but it seems like the author allows anyone to relicense it under GPL or a similar FOSS license (I guess LGPL falls under that category). This would mean that we may choose to use it as either a package, or integrate it fully into RIOT.

http://elm-chan.org/fsw/ff/en/appnote.html#license

**Comment by gebart** [comment]
FAT would be useful for SD card storage, which provide internal wear levelling, but it would probably wear out simpler flash chips too quickly to be useful on those.

**Comment by kaspar030** [comment]
Again I'd like to vote strongly against FAT. We had an implementation in early versions of RIOT. It was a pain in the ass to fix all "special cases" FAT  requires to be usable.

**Comment by authmillenon** [comment]
@kaspar030 FatFS is a full library for FAT for embedded systems. We don't need to do anything, exept providing an interface ;-)

**Comment by authmillenon** [comment]
@gebart I guess `pkg` is the always the best option for external libs, since maintenance in case of updates is easier this way (the avoidance of license hassle is just an extra ;-)).

OT: `embunit` is currently the only non-package module which was included from an external source. Mainly because it IS license-compatible, since it is pretty integral to our testing infrastructure, since we applied a lot of changes by now and to avoid redownloads on Travis.

**Comment by gebart** [comment]
@authmillenon FatFS as a pkg sounds like a good solution. 

btw, do any of the RIOT boards have an implementation for SDHC so far?

**Comment by authmillenon** [comment]
The msba2 has at least a microSD slot, but I don't now if it is SDHC or just SD and if a driver implementation for this exists. Maybe @OlegHahm can shed some light on this.

**Comment by authmillenon** [comment]
@gebart if you like you can adopt this PR, btw. I don't know if I come to continue this in the near future. The interface is inspired mainly by fuse btw ;-)

**Comment by gebart** [comment]
@authmillenon I meant SD host controller interface, sorry about the unclear acronym.

**Comment by gebart** [comment]
I do not have any time to work on this at the moment either, maybe I will adopt this later when things clear up, but for now this one may be up for grabs.

**Comment by authmillenon** [comment]
:+1:

**Comment by waehlisch** [comment]
@x3ro started also to think/work about a file system for RIOT. Maybe he can join the discussion.

**Comment by x3ro** [comment]
Thanks for the ping @waehlisch :) My plan is to be working on bringing SDHC and filesystem support to RIOT over the next couple of months as part of my masters thesis. @authmillenon Would it be an option for me to adopt this PR?

> The msba2 has at least a microSD slot, but I don't now if it is SDHC or just SD and if a driver implementation for this exists. @authmillenon 

As far as I know there is no such driver. I played with the msba2 at the last Hack&Ack to see if I can talk to the SD card but have so far been unsuccessful.

**Comment by authmillenon** [comment]
> @authmillenon Would it be an option for me to adopt this PR?

@x3ro, of course. GitHub has no feature however to just let me give it to you. You have to check out my [corresponding branch](https://github.com/authmillenon/RIOT/tree/filesystem) and open your own PR with it. Then we can close this one.

**Comment by OlegHahm** [comment]
@gebart:
> Contiki has CFS, but that is BSD licensed, so we can't use that code.

Why should this be a problem? AFAIK it shouldn't be a problem to include BSD/MIT/... code into a (L)GPL project - but of course not via versa.

@authmillenon:
> OT: embunit is currently the only non-package module which was included from an external source. 

ccn-lite is also a non-package stemming from an external source - and currently it seems to me that this was a bad idea (although it looked wise back when the decision was originally made). Hence, I would also vote for "packaging" wherever possible.

@gebart @authmillenon @x3ro:
> As far as I know there is no such driver. I played with the msba2 at the last Hack&Ack to see if I can talk to the SD card but have so far been unsuccessful.

I used to work with micro SDHC cards on the MSB-A2 and RIOT (though it was still named µkleos back then) about three years ago, which worked more or less fine (see also https://github.com/RIOT-OS/RIOT/issues/127). However, I haven't tested for a looong time.

**Comment by gebart** [comment]
@OlegHahm 
>Why should this be a problem? AFAIK it shouldn't be a problem to include BSD/MIT/... code into a (L)GPL project - but of course not via versa.

I am sorry, I did not think this through before I wrote it. You are right, AFAIK we are allowed to use BSD sources within an LGPL project such as this. I also found this document regarding different open source licenses and their interoperability:

http://wiki.osdev.org/Licensing

@authmillenon , @OlegHahm What would be a reasonable way to integrate Contiki's CFS into RIOT? 
CFS does not have a separate repo, but is closely integrated into the Contiki repo.

Can we add it as a pkg with some script to fetch a git revision and strip everything except the CFS pieces?

**Comment by OlegHahm** [comment]
I think that could be one way to do this, but probably not the most elegant way. Maybe we should reach out to the Contiki people and see if we can maintain CFS as an external library together on the long run.

**Comment by authmillenon** [comment]
Some thought on file descriptors I had I'd like to share: Maybe any value of `kernel_pid_t` with `> MAXTHREADS` is the best solution, since in POSIX even pthreads have fds and this way we can just have them be the actual PID, while anything else is uniquely identified. This way we have a unique numbering scheme and have the POSIX wrapper for pthreads and file operations as easy as possible. Sockets would also grab from the same pool then (and we have the option to run every socket in it's own light-weight thread (if we'll have something like that) in the future, maybe).

**Comment by gebart** [comment]
@authmillenon watch out for the FDs of stdout, stdin, stderr though. They are defined as 0, 1, and 2 in most implementations.

**Comment by authmillenon** [comment]
Damn you're right. It's even specified this way: http://pubs.opengroup.org/onlinepubs/9699919799/functions/stdin.html

**Comment by x3ro** [comment]
Sorry for my idleness so far :snail: If nobody objects I would open a new pull request from my fork, adding a summary of the discussion so far to that PR. This thread could then be closed and the discussion would continue in the new PR. Does that make sense?

**Comment by gebart** [comment]
@x3ro That makes sense.
Just add the text `#1265` somewhere in the description of your PR and a back-reference will be added inside this thread to your new PR, to make it easier to follow.

**Comment by OlegHahm** [comment]
:+1:

File: sys/fs/fs.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    fs_table[i].mp_len = strlen(mount_point);
+
+    if (fs_table[i].mp_len < 0 || fs_table[i].mp_len > FS_MOUNT_POINT_LEN) {
+        fs_table[i].mp_len = FS_MOUNT_POINT_LEN;
+    }
+
+    fs_table[i].device = device;
+    fs_table[i].type = type;
+
+    int 1;
```
Comment (x3ro) [inline sys/fs/fs.c]: Can anyone explain what this does/should do? :confused:

Comment (gebart) [inline sys/fs/fs.c]: I think they meant to do `return 0;` or `return 1;` or something

**Comment by OlegHahm** [comment]
Should we close this as a memo now?

File: sys/fs/fs.c (hunk 1 above)
Comment (authmillenon) [inline sys/fs/fs.c]: I guess this was meant to be this, too.

**Comment by authmillenon** [comment]
Yes

**Pull request closed without merging**